orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/OperationsController.groovy

```diff
@@ -470,4 +478,58 @@ class OperationsController {
       pipeline.origin = AuthenticatedRequest.spinnakerUserOrigin.orElse('unknown')
     }
   }
+
+  private void addStageAuthorizedRoles(def request, Map pipeline) {
+
+    def jsonObject = jsonSlurper.parseText(request)
+    if (jsonObject.application) {
+      Application application = front50Service.get(jsonObject.application)
+      if (application) {
+        def username = AuthenticatedRequest.getSpinnakerUser().orElse("")
+        def permissions = objectMapper.writeValueAsString(application.getPermission().permissions)
+        def permJsonObject = jsonSlurper.parseText(permissions)
+        def applicationRoles = []
+        permJsonObject.each { item ->
+          List<String> strList = item.getValue()
+          strList.each { item1 ->
+            String[] strArray = item1.split(',')
+            strArray.each { item2 ->
+              if (!applicationRoles.contains(item2.trim())) {
+                applicationRoles.add(item2.trim())
+              }
+            }
+          }
+        }
+        UserPermission.View permission = fiatPermissionEvaluator.getPermission(username);
+        if (permission == null) { // Should never happen?
+          return;
+        }
+        // User has to have all the pipeline roles.
+        Set<Role.View> roleView = permission.getRoles()
+        def userRoles = []
+        roleView.each { it -> userRoles.add(it.getName().trim()) }
+        def stageList = pipeline.stages
+        def stageRoles = []
+        stageList.each { item ->
+          stageRoles = item.selectedStageRoles
+          item.isAuthorized = checkAuthorizedGroups(userRoles, stageRoles, applicationRoles)
+        }
+      }
+    }
+  }
```

Comment:
This entire function feels kind of brittle to me. The fact that we're converting strongly-typed objects into a map and iterating over the keys is quite concerning.

Reply:
I need only the roles, which are of List<String> and it needs to check with stageRoles and applicationRoles. This is the best, we can do.